File: perftest/solver.js
'use strict';

function findPath(maze, start, position) {
    let UP = 1, RIGHT = 2, DOWN = 4, LEFT = 8;

    function str(x, y) {
        return "" + x + "," + y;
    }
    function hasWall(x, y, dir) {
        return maze[y][x] & dir;
    }
    function unvisitedExits(x, y, pathSoFar) {
        let deltas = [[-1, 0, LEFT], [0, -1, UP], [1, 0, RIGHT], [0, 1, DOWN]];
        let result = [];
        for (let i in deltas) {
            let dx = deltas[i][0];
            let dy = deltas[i][1];
            let dir = deltas[i][2];

            let xx = x + dx, yy = y + dy;

            if ((xx >= 0) && (yy >= 0) && (xx < maze[0].length) && (yy < maze.length)
                && !hasWall(x, y, dir) && visited.indexOf(str(xx, yy)) == -1) {
                result.push({x: xx, y: yy, path: pathSoFar.concat([[xx, yy]])});
                visited.push(str(xx, yy));
            }
        }
        return result;
    }

    let target_x = position[0];
    let target_y = position[1];
    let walkers = [{x: start[0], y: start[1], path: [start]}];
    let visited = [str(start)];
    while(walkers.length > 0) {
        let newWalkers = [];
        for (let i in walkers) {
            let walker = walkers[i];
            if ((walker.x == target_x) && (walker.y == target_y)) {
                return walker.path;
            } else {
                let unvisited = unvisitedExits(walker.x, walker.y, walker.path);
                newWalkers = newWalkers.concat(unvisited);
            }
        }
        walkers = newWalkers;
    }
}

module.exports = findPath;

File: src/mazerace/solver.clj
(ns mazerace.solver
  (:require [clojure.tools.logging :as log]
            [clojure.core.async :refer [>! <! go chan close! thread alts! timeout go-loop alt!]]))

(defn has-wall? [maze [x y] direction]
  (let [WALLS {:up 1 :right 2 :down 4 :left 8}]
    (not= 0 (bit-and (get-in maze [y x]) (get WALLS direction)))))

(defn exits [maze [x y]]
  (let [deltas {:left [-1 0] :right [1 0] :down [0 1] :up [0 -1]}
        width (count (first maze))
        height (count maze)]
    (for [[dir [dx dy]] deltas
          :let [xx (+ x dx) yy (+ y dy)]
          :when (and (<= 0 xx (dec width))
                     (<= 0 yy (dec height))
                     (not (has-wall? maze [x y] dir)))]
      [xx yy])))

(defn find-shortest-path [maze from to]
  (log/debug "find-path" from to)
  (letfn [(step [paths visited]
            (mapcat (fn [path]
                      (for [exit (exits maze (last path))
                            :when (not (contains? visited exit))]
                        (conj path exit)))
                    paths))
          (finished? [path] (= (last path) to))]
    (let [paths [[from]]
          visited (hash-set from)]
      (loop [paths paths visited visited]
        (if (empty? paths)
          nil                                               ; no path found
          (let [finished (filter finished? paths)]
            (if (seq finished)
              (first finished)
              (let [paths (step paths visited)]
                (recur paths (into visited (map last paths)))))))))))

(defn start-solver-loop [game send-ch]
  (go
    (<! (timeout 5000))                                     ; pretend to solve the maze
    (loop []
      (let [[_ ch] (alts! [send-ch (timeout 500)])
            path (:path @game)]
        (if (not= ch send-ch)
          (do
            (when (seq path)
              (swap! game assoc :path (rest path))
              (>! send-ch {:move (first path)}))
            (recur))
          (log/debug "Exiting solver loop"))))))

(defn computer-player [[recv-ch send-ch]]
  (let [game (atom {})]
    (start-solver-loop game send-ch)
    (go
      (loop []
        (let [msg (<! recv-ch)]
          (log/debug "Cmp got" msg)
          (if msg
            (do
              (when (:maze msg)
                (reset! game msg))
              (when (:position msg)
                (swap! game assoc
                       :position (:position msg)
                       :path (find-shortest-path (:maze @game) (:position msg) (:target @game))))
              (when (:result game)
                (swap! game assoc :path nil))
              (when (not (:result game))
                (recur)))
            (do
              (log/debug "Closing receive loop")
              (swap! game assoc :path nil)
              (close! recv-ch)
              (close! send-ch))))))))
